nio/client/async_client.py

```diff
@@ -917,6 +919,58 @@ def add_response_callback(
 
         return await self._send(KeysQueryResponse, method, path, data)
 
+    @logged_in
+    async def devices(self) -> Union[DevicesResponse, DevicesError]:
+        """Get the list of devices for the current user.
+
+        Returns either a `DevicesResponse` if the request was successful
+        or a `DevicesError` if there was an error with the request.
+        """
+        method, path = Api.devices(self.access_token)
+
+        return await self._send(DevicesResponse, method, path)
+
+    @logged_in
+    async def delete_devices(
+            self,
+            devices: List[str],
+            auth:    Optional[Dict[str, str]] = None
+    ) -> Union[DeleteDevicesResponse, DeleteDevicesError]:
+        """Delete a list of devices.
+
+        This tells the server to delete the given devices and invalidate their
+        associated access tokens.
+
+        Returns either a `DeleteDevicesResponse` if the request was successful
+        or a `DeleteDevicesError` if there was an error with the request.
+
+        Args:
+            devices (List[str]): A list of devices which will be deleted.
+            auth (Dict): Additional authentication information for
+                the user-interactive authentication API.
+
+        Example:
+            >>> devices = ["QBUAZIFURK", "AUIECTSRND"]
+            >>> auth = {"type": "m.login.password",
+            ...         "user": "example",
+            ...         "password": "hunter1"}
+            >>> await client.delete_devices(devices, auth)
```

Comment:
Does this work that way? As far as i remember you need to do two requests to delete devices. The first one returns a 401 HTTP code and tells you the available auth methods, only then can you provide auth dictionary.

The HttpClient implementation seems to suggest that as well https://github.com/poljar/matrix-nio/blob/master/nio/client/http_client.py#L1252.

We would need a special case for this in `create_matrix_response()` as well. And the docs would need to be updated that it's a two stage thing that first returns a `DeleteDevicesAuthResponse` and only then a `DeleteDevicesResponse`.

Since it isn't a trivial addition anymore I would really like to see some tests for this, unless I'm missing something and we do not need the two stages in fact.

Reply:
It does work for me like this right now, and I'm quite happy about that - I didn't plan to get deep into implementation details. I don't know if we can expect that it works like that, I guess it's probably more clean to do the approach you had in mind.

It's right that in case of missing auth, a DeleteDevicesAuthResponse would be more appropriate, otherwise the DeleteDevicesResponse just says "unknown error".

Further changes concerning two-stage implementation and tests would be out of scope of what I'm willing to do right now. Maybe someone can take this PR as starting point if more work is required

Reply:
Indeed it works, I was still a bit uncomfortable not supporting the auth response since the spec is ambiguous if this should work or not.

Added it myself as well as the tests. Merged, thanks.

Reply:
Great, thank you!